### PR TITLE
feat: DH-21391: Add unit test that highlights edge cases of float vs double parsing

### PR DESCRIPTION
**Pull request by kosak**
This is a subtle topic, but there are a a few edge cases where users will get slightly different floating point answers than expected when using our library. These all boil down to the same reasons:

There are cases where the Java expression `Float.parse(s)` can differ from `(float)Double.parse(s)`.  We use (the equivalent of) the latter everywhere for type inferencing reasons and because our double parser is a third-party library that is very fast.

We are already aware of this, and we provide `FloatStrictParser` for users that want strict behavior that is exactly `Float.parse(s)`.

This issue adds a unit test that exercises `FloatStrictParser` and highlights its differences.

### CODE REVIEW

File: src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        final Object array = result.columns()[0].data();
+        final Object element0 = Array.get(array, 0);
+        boolean isSame = element0.equals(expectedValue);
+
+        if (isSame) {
+            Assertions.assertThat(fidelity).isEqualTo(ParseFidelity.SAME);
+        } else {
+            Assertions.assertThat(fidelity).isEqualTo(ParseFidelity.DIFFERENT);
+        }
```
Comment (devinrsmith) [inline src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java]: This seems like we are mixing up what is expected and what is the result; I would say we should switch on fidelity, and then have an appropriate assertion:

```java
if (fidelity == SAME) {
  assertThat(element0).isEqualTo(expectedValue);
} else if (fidelity == DIFFERENT) {
  assertThat(element0).isNotEqualTo(expectedValue);
}  else {
  throw new IllegalStateException();
}
```

File: src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        final CsvSpecs specs = CsvTestUtil.defaultCsvBuilder().parsers(Collections.singletonList(parser)).build();
+        final InputStream stream = CsvTestUtil.toInputStream(source);
+        CsvReader.Result result;
+        try {
+            result = CsvReader.read(specs, stream, CsvTestUtil.makeMySinkFactory());
+        } catch (CsvReaderException e) {
+            Assertions.assertThat(fidelity).isEqualTo(ParseFidelity.PARSE_FAIL);
+            return;
+        }
+
```
Comment (devinrsmith) [inline src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java]: If you expected to fail, but didn't, you should fail.

File: src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        } catch (CsvReaderException e) {
+            failed = true;
+        }
+
+        if (fidelity == ParseFidelity.PARSE_FAIL) {
+            Assertions.assertThat(failed);
+            return;
+        }
+
+        Assertions.assertThat(!failed);
```
Comment (devinrsmith) [inline src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java]: Similarly:

```java
assertThat(failed).isFalse();
```

File: src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        assert result != null; // make linter happy
```
Comment (devinrsmith) [inline src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java]: Is linter smart enough to know this implies not null?

```java
assertThat(result).isNotNull();
```

File: src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        CsvReader.Result result = null;
+        boolean failed = false; // optimistically assume success
+        try {
+            result = CsvReader.read(specs, stream, CsvTestUtil.makeMySinkFactory());
+        } catch (CsvReaderException e) {
+            failed = true;
+        }
+
+        if (fidelity == ParseFidelity.PARSE_FAIL) {
+            Assertions.assertThat(failed);
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java]: Wow, thank you for catching that. I'm kind of embarrassed. PTAL

File: src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java (hunk 4 above)
Comment (kosak) [reply src/test/java/io/deephaven/csv/testutil/ParserFidelityTest.java]: Don't need it any more because of the above fixes. Apparently the linter was alerting me for real but I was ignoring it.
